### DISCUSSION TITLE
"packaged" docker image for step vs. build action every run

### DISCUSSION THREAD

**Discussion by ckcr4lyf** [community/community#44956] [Actions] [answered]
### Select Topic Area

General

### Body

Hey guys, I was wondering for best practices with Github actions.

Currently I made an action which is defined via `action,yml` and a Dockerfile (see: https://github.com/ckcr4lyf/actions-kiryuu-qbit-test) . However, it seems GH Actions build this image on every run, which I feel is wasteful (https://github.com/ckcr4lyf/kiryuu/actions/runs/3973519644/jobs/6812227636#step:2:1)

For situations like this, where the image does not need to "be updated" on every run, is it better to have it built into a docker image and then make the step use the image directly?

**Comment by jge162** [comment, marked as answer]
@ckcr4lyf One way to use a pre-built image is to push the image to a container registry, such as
 
Docker Hub, and then use the uses syntax in the workflow file to reference the image.

This way, the image is only built when you update it and pushed to the container registry.

**Reply by ckcr4lyf** [reply]
Thanks, I am able to use pre built images, I was just wondering what the "better" approach is, use a pre-built, or have GH actions build it for every run.

Seems the default to publish a docker actions is have it build every time

**Reply by jge162** [reply]
building the image every time on GitHub Actions can ensure that you always have the latest versions of dependencies and security patches, but it can also increase the time and resource usage.

It depends on your use case and resources, you can use a pre-built image and update them regularly or you can use GitHub Actions to build the image every time.

**Reply by ckcr4lyf** [reply]
Got it, there are tradeoffs on either side. Thanks for explaining, I will mark it as solved

**Reply by OR13** [reply]
> then use the uses syntax in the workflow file to reference the image.

It would be excellent to provide an example of this syntax... 

Here:

https://docs.github.com/en/actions/creating-actions/creating-a-docker-container-action#creating-an-action-metadata-file

And...

In the hello world repo, add a workflow file that uses both the "build each time" and the "use existing image" path, for example:

https://github.com/actions/hello-world-docker-action/blob/main/action.yml#L13
